File: 4thPractical/SquareRoot.cc
#include <cstdlib>
#include <iostream>

using namespace std;

double f(int c, double x); 
void updateDomain(double &a, double &b, int c, double p);
double squareRoot(int c, double tol);

int main(int argc, char* argv[])
{
    int N = 10; 
    double tol = 1.0e-14;

    if(argc > 1)
    {
        N = atoi(argv[1]);
    }
    if(argc > 2)
    {
        tol = atof(argv[2]);
    }

    while(N < 2 | N > 10000)
    {
        cout << "N: ";
        cin >> N; 
    }

    printf("x\tsqrt(x)\t\trel error\n");
    for(int n = 2; n <= N; n++)
    {
        double ans = squareRoot(n, tol);
        double err = (n - ans*ans) / n;
        if (err < 0.0)
        {
            err *= -1.0;
        }
        printf("%d\t%.5f\t\t%.2e\n", n, ans, err);
    }

    return 0;
}

double f(int c, double x)
{
    return c - x * x;
}

void updateDomain(double &a, double &b, int c, double p)
{
    if(f(c, a) * f(c, p) < 0)
    {
        b = p;
    }
    else
    {
        a = p;
    }
}

double squareRoot(int c, double tol)
{
    double a = 1, b = 100, p, err; 
    
    do
    {
        p = (a + b) / 2.0; 
        err = (b - a) / 2.0; 
        updateDomain(a, b, c, p);
    } 
    while(err > tol);

    return p;
}

File: 4thPractical/SquareRoot.py
import sys

def f(c, x):
    return c - x**2


def update_domain(a, b, c, p):
    if f(c, a) * f(c, p) < 0:
        return [a, p]
    else:
        return [p, b]


def square_root(c, tol):
    a = 1
    b = 100
    err = 1
    while err > tol:
        p = (a + b) / 2
        err = (b - a) / 2
        [a, b] = update_domain(a, b, c, p)
    return p


N = 10
tol = 1e-14

if len(sys.argv) > 1:
    N = int(sys.argv[1])
if len(sys.argv) > 2:
    tol = float(sys.argv[1])

while not 2 <= N <= 10000:
    N = int(input("N: "))

print('x\tsqrt(x)\t\trel error')
for n in range(2, N+1):
    ans = square_root(n, tol)
    err = abs(n - ans*ans) / n
    print('%d\t%.5f\t\t%.2e' % (n, ans, err))
